### PR TITLE
fuzzing tests added for APLN RFC compliance

### PR DESCRIPTION
**Pull request by 7echno**
depends on tomato42/tlslite-ng#150

obsoletes #92

### CODE REVIEW

File: scripts/test-alpn-negotiation.py [hunk 1]
```diff
@@ -193,6 +194,48 @@ def main():
     node.add_child(ExpectClose())
     conversations["empty extension"] = conversation
 
+    # invalid length test
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: please tell which length, same for the name of the test

File: scripts/test-alpn-negotiation.py [hunk 2]
```diff
@@ -193,6 +194,48 @@ def main():
...
+    conversation = Connect(host, port)
+    node = conversation
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA,
+               CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV]
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.1'),
+                                                       bytearray(b'http/2')])}
+    msg = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    node = node.add_child(fuzz_message(msg, substitutions={-17: 9}))
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: please add a comment stating why -17 and why 9

File: scripts/test-alpn-negotiation.py [hunk 3]
```diff
@@ -193,6 +194,48 @@ def main():
...
+    node = node.add_child(ExpectAlert(AlertLevel.fatal,
+                                      AlertDescription.decode_error))
+    node = node.add_child(ExpectClose())
+    conversations["fuzz length under"] = conversation
+
+    # invalid length test
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: again, which length

File: scripts/test-alpn-negotiation.py [hunk 4]
```diff
@@ -193,6 +194,48 @@ def main():
...
+    conversation = Connect(host, port)
+    node = conversation
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA,
+               CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV]
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.1'),
+                                                       bytearray(b'http/2')])}
+    msg = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    node = node.add_child(fuzz_message(msg, substitutions={-7: 8}))
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: please add comment explaining why -7 and why 8

File: scripts/test-alpn-negotiation.py [hunk 5]
```diff
@@ -193,6 +194,54 @@ def main():
...
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.1'),
+                                                       bytearray(b'http/2')])}
+    msg = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    # -17 is position of second byte in 2 byte long lenght of "protocol_name_list"
+    # setting it to value od 9 ( bytes ) will hide the second item in the "protocol_name_list"    
+    node = node.add_child(fuzz_message(msg, substitutions={-17: 9}))
+    node = node.add_child(ExpectAlert(AlertLevel.fatal,
+                                      AlertDescription.decode_error))
+    node = node.add_child(ExpectClose())
+    conversations["underflow lenght of protocol_name_list"] = conversation
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: typo "lenght"

File: scripts/test-alpn-negotiation.py [hunk 6]
```diff
@@ -193,6 +194,54 @@ def main():
...
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.1'),
+                                                       bytearray(b'http/2')])}
+    msg = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    # -17 is position of second byte in 2 byte long lenght of "protocol_name_list"
+    # setting it to value of 18 ( bytes ) will raise the lenght value for 2 more bytes    
+    node = node.add_child(fuzz_message(msg, substitutions={-17: 18}))
+    node = node.add_child(ExpectAlert(AlertLevel.fatal,
+                                      AlertDescription.decode_error))
+    node = node.add_child(ExpectClose())
+    conversations["overflow lenght of protocol_name_list"] = conversation
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: typo "lenght"

File: scripts/test-alpn-negotiation.py [hunk 7]
```diff
@@ -193,6 +194,54 @@ def main():
...
+
+    # overflow lenght of "protocol_name_list" test
+    conversation = Connect(host, port)
+    node = conversation
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA,
+               CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV]
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.1'),
+                                                       bytearray(b'http/2')])}
+    msg = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    # -17 is position of second byte in 2 byte long lenght of "protocol_name_list"
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: typo "lenght"

File: scripts/test-alpn-negotiation.py [hunk 8]
```diff
@@ -193,6 +194,54 @@ def main():
...
 
+    # underflow length of "protocol_name_list" test
+    conversation = Connect(host, port)
+    node = conversation
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA,
+               CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV]
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.1'),
+                                                       bytearray(b'http/2')])}
+    msg = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    # -17 is position of second byte in 2 byte long lenght of "protocol_name_list"
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: typo "lenght"

File: scripts/test-alpn-negotiation.py [hunk 9]
```diff
@@ -193,6 +194,54 @@ def main():
...
+    # setting it to value od 9 ( bytes ) will hide the second item in the "protocol_name_list"
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: typo "od"

Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: there should be no spaces inside parentheses

File: scripts/test-alpn-negotiation.py [hunk 10]
```diff
@@ -193,6 +194,54 @@ def main():
...
+
+    # overflow lenght of last item in "protocol_name_list" test
+    conversation = Connect(host, port)
+    node = conversation
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA,
+               CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV]
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.1'),
+                                                       bytearray(b'http/2')])}
+    msg = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    # -7 is position of a lenght ( 1 byte long ) for the last item in "protocol_name_list"
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: typo "lenght"

**Comment by 7echno** [comment]
@tomato42  Can you please check why travis-ci was not successful ?

**Comment by tomato42** [comment]
test-alpn-negotiation.py when run against tlslite-ng failed

**Comment by 7echno** [comment]
@tomato42 Seems good to go now.

File: tests/tlslite-ng.json [hunk 11]
```diff
@@ -8,6 +8,40 @@
      "environment": {"PYTHONPATH": "."},
      "tests" : [
          {"name" : "test-aes-gcm-nonces.py" },
+         {"name" : "test-alpn-negotiation.py"
```
Comment (tomato42) [inline tests/tlslite-ng.json]: comma missing at the end of line

File: scripts/test-alpn-negotiation.py (hunk 5 above)
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: typo not fixed

File: scripts/test-alpn-negotiation.py (hunk 7 above)
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: typo not fixed

File: scripts/test-alpn-negotiation.py (hunk 6 above)
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: typo

File: scripts/test-alpn-negotiation.py (hunk 10 above)
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: typo

File: scripts/test-alpn-negotiation.py [hunk 12]
```diff
@@ -390,7 +390,7 @@ def main():
                                          AlertDescription.close_notify))
     node = node.add_child(ExpectAlert())
     node.next_sibling = ExpectClose()
-    conversations["16269 size of array"] = conversation
+    conversations["16269 byte long array"] = conversation
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: that changes don't match the commit description

File: scripts/test-alpn-negotiation.py [hunk 13]
```diff
@@ -210,32 +210,32 @@ def main():
     node = node.add_child(ExpectClose())
     conversations["underflow length of protocol_name_list"] = conversation
 
-    # overflow lenght of "protocol_name_list" test
+    # overflow length of "protocol_name_list" test
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: squash to fixes with the commits in which they were introduced

File: scripts/test-alpn-negotiation.py [hunk 14]
```diff
@@ -360,7 +359,150 @@ def main():
...
+    node = node.add_child(Connect(host, port))
+    close.add_child(node)
+
+    node = node.add_child(ResetHandshakeHashes())
+    node = node.add_child(ResetRenegotiationInfo())
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.1')]),
+	   ExtensionType.renegotiation_info:None}
+    node = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    ext = {ExtensionType.renegotiation_info:None,
+           ExtensionType.alpn: lambda _, ext: ext == ALPNExtension().create([bytearray(b'http/1.1')])}
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: no need to use lambdas any more if you don't use the session information

`ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.1')])` will have exact same effect

File: scripts/test-alpn-negotiation.py [hunk 15]
```diff
@@ -360,7 +359,150 @@ def main():
...
+    # resumption with alpn change
+
+    conversation = Connect(host, port)
+    node = conversation
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA]
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.1')]),
+	   ExtensionType.renegotiation_info:None}
+    node = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    ext = {ExtensionType.renegotiation_info: None,
+           ExtensionType.alpn: lambda _, ext: ext == ALPNExtension().create([bytearray(b'http/1.1')])}
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: same here, no need for lambdas

File: scripts/test-alpn-negotiation.py [hunk 16]
```diff
@@ -360,7 +359,150 @@ def main():
...
+    node = node.add_child(Connect(host, port))
+    close.add_child(node)
+
+    node = node.add_child(ResetHandshakeHashes())
+    node = node.add_child(ResetRenegotiationInfo())
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'h2')]),
+	   ExtensionType.renegotiation_info:None}
+    node = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    ext = {ExtensionType.renegotiation_info:None,
+           ExtensionType.alpn: lambda _, ext: ext == ALPNExtension().create([bytearray(b'h2')])}
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: no need for lambda

File: scripts/test-alpn-negotiation.py [hunk 17]
```diff
@@ -360,7 +359,150 @@ def main():
...
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: no need for lambda

File: scripts/test-alpn-negotiation.py [hunk 18]
```diff
@@ -0,0 +1,582 @@
...
+    conversations["sanity"] = conversation
+
+    conversation = Connect(host, port)
+    node = conversation
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA,
+               CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV]
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.1')])}
+    node = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    ext = {ExtensionType.renegotiation_info: None,
+           ExtensionType.alpn: lambda _, ext: ext == ALPNExtension().create([bytearray(b'http/1.1')])}
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: no need for lambda

File: scripts/test-alpn-negotiation.py [hunk 19]
```diff
@@ -0,0 +1,582 @@
...
+
+    conversation = Connect(host, port)
+    node = conversation
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA,
+               CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV]
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.'),
+                                                       bytearray(b'http/1.1')])}
+    node = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    ext = {ExtensionType.renegotiation_info: None,
+           ExtensionType.alpn: lambda _, ext: ext == ALPNExtension().create([bytearray(b'http/1.1')])}
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: no need for lambda

File: scripts/test-alpn-negotiation.py (hunk 14 above)
Comment (7echno) [inline scripts/test-alpn-negotiation.py]: @tomato42 TypeError: 'ALPNExtension' object is not callable
Is there anything I need to add/import for this to work ?

File: scripts/test-alpn-negotiation.py [hunk 20]
```diff
@@ -0,0 +1,582 @@
...
+    conversation = Connect(host, port)
+    node = conversation
+    ciphers = [CipherSuite.TLS_RSA_WITH_AES_128_CBC_SHA,
+               CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV]
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'http/1.X')])}
+    node = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    node = node.add_child(ExpectAlert(AlertLevel.fatal,
+                                      AlertDescription.no_application_protocol))
+    node.add_child(ExpectClose())
+    conversations["only http/1.1X"] = conversation
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: that description doesn't match what the client is sending

File: scripts/test-alpn-negotiation.py [hunk 21]
```diff
@@ -0,0 +1,582 @@
...
+    close = ExpectClose()
+    node.next_sibling = close
+    node = node.add_child(ExpectClose())
+    node = node.add_child(Close())
+    node = node.add_child(Connect(host, port))
+    close.add_child(node)
+
+    node = node.add_child(ResetHandshakeHashes())
+    node = node.add_child(ResetRenegotiationInfo())
+    ext = {ExtensionType.alpn: ALPNExtension().create([bytearray(b'h2')]),
```
Comment (tomato42) [inline scripts/test-alpn-negotiation.py]: wouldn't it be better to expect http1.1 here, given that neither tlslite nor openssl -www support http 2 it makes the usefulness of the test case rather limited

File: tests/tlslite-ng.json [hunk 22]
```diff
@@ -8,6 +8,40 @@
...
+          "exp_pass": false
+         },
+         {"name" : "test-alpn-negotiation.py",
+          "arguments" : ["renegotiation without protocol change"],
+          "comment": "tlslite-ng doesn't support renegotiation",
+          "exp_pass": false
+         },
+         {"name" : "test-alpn-negotiation.py",
+          "arguments" : ["resumption with alpn"],
+          "comment": "only session id based resumption is supported",
```
Comment (tomato42) [inline tests/tlslite-ng.json]: but the test case is using session id based resumption...

It looks to me like the issue is with the tlslite-ng not supporting HTTP/2.0...

File: tests/tlslite-ng.json [hunk 23]
```diff
@@ -8,6 +8,40 @@
...
+          "exp_pass": false
+         },
+         {"name" : "test-alpn-negotiation.py",
+          "arguments" : ["resumption with alpn change"],
+          "comment": "only session id based resumption is supported",
```
Comment (tomato42) [inline tests/tlslite-ng.json]: same here

File: scripts/test-alpn-negotiation.py (hunk 20 above)
Comment (7echno) [inline scripts/test-alpn-negotiation.py]: @tomato42 Should be fixed now.

File: scripts/test-alpn-negotiation.py (hunk 21 above)
Comment (7echno) [inline scripts/test-alpn-negotiation.py]: @tomato42 Fixed.

File: tests/tlslite-ng.json (hunk 22 above)
Comment (7echno) [inline tests/tlslite-ng.json]: @tomato42 Fixed in test, comment should fit now.

File: tests/tlslite-ng.json (hunk 23 above)
Comment (7echno) [inline tests/tlslite-ng.json]: @tomato42 Comment fixed.

**Comment by 7echno** [comment]
@tomato42 Seems fine now, can you check it ?

**Comment by 7echno** [comment]
@tomato42 Seems fine now, can you check it ?

**Comment by tomato42** [comment]
LGTM. Thanks!

**Pull request merged**